File: app/src/main/java/org/checklist/comics/comicschecklist/parser/Parser.java
package org.checklist.comics.comicschecklist.parser;

import org.checklist.comics.comicschecklist.database.entity.ComicEntity;
import org.checklist.comics.comicschecklist.log.CCLogger;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Locale;

/**
 * Abstract class which give instruction how sub-class should parse data from HTML pages.
 */
public abstract class Parser {

    private static final String TAG = Parser.class.getSimpleName();

    public abstract ArrayList<ComicEntity> initParser();
    public abstract ArrayList<ComicEntity> parseUrl(String url);
    public abstract String searchTitle(Object object);
    public abstract String searchReleaseDate(Object object);
    public abstract String searchDescription(Object object);
    public abstract String searchCover(Object object);
    public abstract String searchFeature(Object object);
    public abstract String searchPrice(Object object);

    /**
     * Elaborate a date, in {@link Date} format, from given date which is in {@link String} format.
     * @param releaseDate the date to elaborate
     * @return the date in another format
     */
    Date elaborateDate(String releaseDate) {
        CCLogger.v(TAG, "elaborateDate - start");
        Date date = null;
        try {
            date = new SimpleDateFormat("dd/MM/yyyy", Locale.getDefault()).parse(releaseDate);
        } catch (ParseException e) {
            CCLogger.w(TAG, "elaborateDate - Error while elaborating Date from " + releaseDate + " " + e.toString());
        } finally {
            if (date == null) {
                date = new Date();
                date.setTime(System.currentTimeMillis());
                CCLogger.d(TAG, "elaborateDate - Setting date from current time");
            }
        }
        CCLogger.v(TAG, "elaborateDate - end - " + date.toString());
        return date;
    }

}

File: app/src/main/java/org/checklist/comics/comicschecklist/parser/Parser.kt
package org.checklist.comics.comicschecklist.parser

import org.checklist.comics.comicschecklist.database.entity.ComicEntity
import org.checklist.comics.comicschecklist.log.CCLogger

import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.ArrayList
import java.util.Date
import java.util.Locale

/**
 * Abstract class which give instruction how sub-class should parse data from HTML pages.
 */
abstract class Parser {

    abstract fun initParser(): ArrayList<ComicEntity>
    abstract fun parseUrl(url: String): ArrayList<ComicEntity>
    abstract fun searchTitle(`object`: Any): String?
    abstract fun searchReleaseDate(`object`: Any): String?
    abstract fun searchDescription(`object`: Any): String?
    abstract fun searchCover(`object`: Any): String?
    abstract fun searchFeature(`object`: Any): String?
    abstract fun searchPrice(`object`: Any): String?

    /**
     * Elaborate a date, in [Date] format, from given date which is in [String] format.
     * @param releaseDate the date to elaborate
     * @return the date in another format
     */
    internal fun elaborateDate(releaseDate: String): Date {
        CCLogger.v(TAG, "elaborateDate - start")
        var date: Date? = null
        try {
            date = SimpleDateFormat("dd/MM/yyyy", Locale.getDefault()).parse(releaseDate)
        } catch (e: ParseException) {
            CCLogger.w(TAG, "elaborateDate - Error while elaborating Date from " + releaseDate + " " + e.toString())
        } finally {
            if (date == null) {
                date = Date()
                date.time = System.currentTimeMillis()
                CCLogger.d(TAG, "elaborateDate - Setting date from current time")
            }
        }
        CCLogger.v(TAG, "elaborateDate - end - " + date.toString())
        return date
    }

    companion object {

        private val TAG = Parser::class.java.simpleName
    }

}
